internal/broadcast/manager.go

```diff
@@ -144,21 +150,32 @@ func (bm *broadcastManager) dispatchBatch(ctx context.Context, state *batch.Disp
 	return bm.batchpin.SubmitPinnedBatch(ctx, &state.Persisted, state.Pins)
 }
 
-func (bm *broadcastManager) publishBlobs(ctx context.Context, newMsg *data.NewMessage) error {
-	for _, d := range newMsg.ResolvedData.DataToPublish {
-		// Stream from the local data exchange ...
-		reader, err := bm.exchange.DownloadBLOB(ctx, d.Blob.PayloadRef)
-		if err != nil {
-			return i18n.WrapError(ctx, err, i18n.MsgDownloadBlobFailed, d.Blob.PayloadRef)
-		}
-		defer reader.Close()
-
-		// ... to the shared storage
-		d.Data.Blob.Public, err = bm.sharedstorage.PublishData(ctx, reader)
-		if err != nil {
-			return err
+func (bm *broadcastManager) publishBlobs(ctx context.Context, data fftypes.DataArray) error {
+	for _, d := range data {
+		// We only need to send a blob if there is one, and it's not been uploaded to the shared storage
+		if d.Blob != nil && d.Blob.Hash != nil && d.Blob.Public == "" {
+			blob, err := bm.database.GetBlobMatchingHash(ctx, d.Blob.Hash)
+			if err != nil {
+				return err
+			}
+			if blob == nil {
+				return i18n.NewError(ctx, i18n.MsgBlobNotFound, d.Blob.Hash)
+			}
+
+			// Stream from the local data exchange ...
+			reader, err := bm.exchange.DownloadBLOB(ctx, blob.PayloadRef)
+			if err != nil {
+				return i18n.WrapError(ctx, err, i18n.MsgDownloadBlobFailed, blob.PayloadRef)
+			}
+			defer reader.Close()
+
+			// ... to the shared storage
+			d.Blob.Public, err = bm.sharedstorage.PublishData(ctx, reader)
```

Comment:
This was the part where I got stuck before. How does this public ref 1) get stored in the database, and 2) not cause the data hash to change? Hasn't the batch been sealed and stored before we reach here?

Reply:
After looking further, it looks like the answer to (2) is that the hash calculation changed to not include the public ref as part of the hash. That's good.

Still can't figure out (1), ie where this data item is updated in the database with the public reference received here. It's definitely updated when it's received again in the event aggregator, but I'd expect it's updated prior to that as well... I'm likely just missing something.

Reply:
Great point. It's _not_ currently at the point the batch is written to the comms layer. As you observe, it is when it comes back in the other end at the end of this PR - but I actually optimized that away in a later change.

I will add it back in, as it's only a performance hit for blobs. But if it's ok will do that as a separate PR.

Reply:
Added in https://github.com/hyperledger/firefly/pull/599/commits/17b9f907be4877ef14e7384e4bc4a1aa1a34a6e9 in #599